File: TOANDFRO.scala
import scala.io.StdIn

object TOANDFRO {
  def main(args: Array[String]): Unit = {
    var eof = false

    while(!eof) {
      val columnCount = StdIn.readLine.toInt
      if (columnCount == 0) eof = true
      else {
        val encryptedMessage = StdIn.readLine
        val parts: List[String] = createParts(encryptedMessage, columnCount)
        val reversed: List[String] = reverseAlternate(parts)
        val decryptedMessage: String = decryptList(reversed)
        println(decryptedMessage)
      }
    }
  }

  def createParts(s: String, l: Int): List[String] = {
    if (s.length > l) {
      val parts = s.splitAt(l)
      parts._1 :: createParts(parts._2, l)
    }
    else s :: List()
  }

  def reverseAlternate(l: List[String]): List[String] = {
    // There's this easy way of doing list operation, but as -
    // of now I am here to learn
    // for(i <- 0 until l.length) { }
    reverseAlternateR(l.head, l.tail.head, l.tail.tail)
  }

  def reverseAlternateR(first: String, second: String, tail: List[String]): List[String] = {
    if (tail.isEmpty ) {
      first :: second.reverse :: List()
    } else if (tail.length == 1) {
      first :: second.reverse :: tail
    }
    else {
      first :: second.reverse :: reverseAlternateR(tail.head, tail.tail.head, tail.tail.tail)
    }
  }

  def decryptList(l: List[String]): String = {
    var currentCount = 0
    val columnCount = l(0).length // These are padded to be equal always
    val message = new StringBuilder()
    for (column <- 0 until columnCount) {
      l.foreach((s) => { message.append(s(column)) })
    }
    message.toString
  }
}

File: toandfro.py
from sys import stdin

def create_parts(s, l):
  while(s):
    yield list(s[:l])
    s = s[l:]

def reverse_alternate(s, l):
  reverse = []
  for (i, parts) in enumerate(create_parts(s, l)):
    if i % 2 != 0: 
      parts.reverse()
    reverse.append(parts)
  return reverse

def decrypt_list(ll):
  column_count = len(ll[0])
  message_list = []
  for i in range(column_count):
    for l in ll:
      message_list.append(l[i])
  decrypted = ''.join(message_list)
  return decrypted


def main():
  eof = False
  while not eof:
    column_count = int(stdin.readline())
    if column_count == 0:
      eof = True
    else:
      encrypted_message = stdin.readline().strip()
      ll = reverse_alternate(encrypted_message, column_count)
      decrypted = decrypt_list(ll)
      print decrypted

if __name__ == '__main__':
  main()
